File: front/src/lib/order-amount.ts
import { D } from "./decimal";
import { FEE_RATE, KR_TAX_RATE, US_TAX_MIN_USD, US_TAX_RATE } from "./mock-data";

/**
 * 매수/매도 금액 미리보기 계산. `back/src/main/java/com/baedang/trading/service/OrderAmountCalculator.java`
 * (김다훈님 구현)와 같은 규칙을 프론트에서 미리보기용으로 재현한 것입니다 — 실제 체결 금액은
 * 항상 백엔드가 다시 계산해서 확정합니다. 여기 값은 화면 미리보기 전용입니다.
 *
 * <p>국내: gross = qty * price, fee = round(gross * FEE_RATE), 매도세 = round(gross * KR_TAX_RATE)
 * 미국: 주당 USD 가격을 센트 반올림한 뒤 수량·환율을 적용하고, 원화 거래대금에서 수수료 계산
 *
 * <p>전부 {@link D}(decimal.js)로 계산합니다 — 순수 `number` 곱셈은 이진 부동소수점
 * 오차가 생겨서 수수료·세금처럼 반올림 규칙이 정확해야 하는 금액 계산엔 부적합합니다.
 */
export type OrderAmount = {
  grossAmount: number; // KRW
  fee: number; // KRW
  tax: number; // KRW
  netAmount: number; // KRW — 매수는 차감액, 매도는 입금액
};

export function calculateOrderAmount(params: {
  side: "매수" | "매도";
  quantity: number;
  /** 백엔드 응답은 금액을 문자열로 내려주므로(정밀도 보존) 문자열도 그대로 받는다. */
  price: number | string;
  currency: "KRW" | "USD";
  usdKrwRate: number;
}): OrderAmount {
  const { side, quantity, price, currency, usdKrwRate } = params;
  const qty = new D(quantity);
  const unitPrice = new D(price);

  if (currency === "KRW") {
    const grossAmount = qty.times(unitPrice).toDecimalPlaces(0);
    const fee = grossAmount.times(FEE_RATE).toDecimalPlaces(0);
    const tax = side === "매도" ? grossAmount.times(KR_TAX_RATE).toDecimalPlaces(0) : new D(0);
    const netAmount = side === "매수" ? grossAmount.plus(fee) : grossAmount.minus(fee).minus(tax);
    return {
      grossAmount: grossAmount.toNumber(),
      fee: fee.toNumber(),
      tax: tax.toNumber(),
      netAmount: netAmount.toNumber(),
    };
  }

  // 미국 — 백엔드와 동일하게 주당 가격을 센트로 먼저 확정한 뒤 수량과 환율을 적용한다.
  const rate = new D(usdKrwRate);
  const roundedUnitPrice = unitPrice.toDecimalPlaces(2);
  const grossUsd = roundedUnitPrice.times(qty);
  const grossAmount = grossUsd.times(rate).toDecimalPlaces(0);
  const fee = grossAmount.times(FEE_RATE).toDecimalPlaces(0);

  let tax = new D(0);
  if (side === "매도") {
    const secFeeUsd = D.max(grossUsd.times(US_TAX_RATE), US_TAX_MIN_USD).toDecimalPlaces(2);
    tax = secFeeUsd.times(rate).toDecimalPlaces(0);
  }

  const netAmount = side === "매수" ? grossAmount.plus(fee) : grossAmount.minus(fee).minus(tax);
  return {
    grossAmount: grossAmount.toNumber(),
    fee: fee.toNumber(),
    tax: tax.toNumber(),
    netAmount: netAmount.toNumber(),
  };
}

/**
 * 이 예산(`availableCash`, 원)으로 매수할 수 있는 최대 수량. 수수료 공식을 여기서
 * 다시 구현하지 않고, 나눗셈으로 대략 추정한 뒤 {@link calculateOrderAmount}로
 * ±1주 오차를 보정한다 — 원 단위 반올림 때문에 나눗셈만으로는 경계값에서 한 주
 * 어긋날 수 있어서다. 매수 화면에서 "얼마까지 살 수 있는지" 입력 상한을 정할 때
 * 쓴다(매도는 보유 수량이 상한이라 이 함수가 필요 없다).
 */
export function maxAffordableQuantity(params: {
  price: number | string;
  currency: "KRW" | "USD";
  usdKrwRate: number;
  availableCash: number;
}): number {
  const { price, currency, usdKrwRate, availableCash } = params;
  const unitPrice = new D(price);
  if (availableCash <= 0 || unitPrice.lessThanOrEqualTo(0)) return 0;

  const unitCostKrw = currency === "USD" ? unitPrice.times(usdKrwRate) : unitPrice;
  const estimatedCostPerShare = unitCostKrw.times(new D(1).plus(FEE_RATE));
  let qty = Math.max(0, new D(availableCash).dividedBy(estimatedCostPerShare).floor().toNumber());

  const affordable = (q: number) =>
    calculateOrderAmount({ side: "매수", quantity: q, price, currency, usdKrwRate }).netAmount <= availableCash;

  // 경계 보정은 설계상 ±1~2주 수준이라 실무에선 반복이 1~2회로 끝나지만,
  // (코드 리뷰, PR #124, SOL4R1S님) 혹시 모를 비정상 입력값으로 추정값이
  // 크게 어긋나는 최악의 경우에도 calculateOrderAmount를 무한정 다시 호출하지
  // 않도록 반복 횟수를 제한한다 — 넘으면 그 시점까지 보정한 값을 그대로 쓴다.
  const MAX_CORRECTION_STEPS = 10;
  for (let steps = 0; qty > 0 && !affordable(qty) && steps < MAX_CORRECTION_STEPS; steps++) qty--;
  for (let steps = 0; affordable(qty + 1) && steps < MAX_CORRECTION_STEPS; steps++) qty++;
  return qty;
}

File: investup_app/lib/features/stock_detail/order_amount.dart
import 'package:decimal/decimal.dart';

/// 주문 금액 미리보기 계산 — `front/src/lib/order-amount.ts`와 같은 규칙을
/// Dart로 옮긴 것. 실제 체결 금액은 항상 백엔드가 다시 계산해서 확정하며
/// 여기 값은 화면 미리보기 전용이다.
///
/// 국내: gross = qty * price, fee = round(gross * 0.01%), 매도세 = round(gross * 0.2%)
/// 미국: 주당 USD 가격을 센트 반올림한 뒤 수량·환율을 적용하고 원화에서 수수료 계산
final _feeRate = Decimal.parse('0.0001'); // 0.01%, 매수·매도 공통
final _krTaxRate = Decimal.parse('0.002'); // 국내 매도세 0.2%
final _usTaxRate = Decimal.parse('0.0000206'); // 미국 SEC Fee
final _usTaxMinUsd = Decimal.parse('0.01');

/// 미리보기 금액 묶음 — 전부 원화 정수.
class OrderAmount {
  const OrderAmount({
    required this.grossAmount,
    required this.fee,
    required this.tax,
    required this.netAmount,
  });

  final Decimal grossAmount;
  final Decimal fee;
  final Decimal tax;

  /// 매수는 차감액, 매도는 입금액.
  final Decimal netAmount;
}

Decimal _max(Decimal a, Decimal b) => a >= b ? a : b;

/// 매수/매도 금액 미리보기. [price]는 종목 통화 기준 주당 가격.
OrderAmount calculateOrderAmount({
  required String side, // 'BUY' | 'SELL'
  required int quantity,
  required Decimal price,
  required String currency, // 'KRW' | 'USD'
  required Decimal usdKrwRate,
}) {
  final qty = Decimal.parse('$quantity');
  final buy = side == 'BUY';

  if (currency == 'KRW') {
    final gross = (qty * price).round(scale: 0);
    final fee = (gross * _feeRate).round(scale: 0);
    final tax = buy ? Decimal.zero : (gross * _krTaxRate).round(scale: 0);
    final net = buy ? gross + fee : gross - fee - tax;
    return OrderAmount(
      grossAmount: gross,
      fee: fee,
      tax: tax,
      netAmount: net,
    );
  }

  // 미국 — 백엔드와 동일하게 주당 가격을 센트로 먼저 확정한 뒤 수량과 환율을 적용한다.
  final roundedUnitPrice = price.round(scale: 2);
  final grossUsd = roundedUnitPrice * qty;
  final gross = (grossUsd * usdKrwRate).round(scale: 0);
  final fee = (gross * _feeRate).round(scale: 0);

  var tax = Decimal.zero;
  if (!buy) {
    final secFeeUsd = _max(grossUsd * _usTaxRate, _usTaxMinUsd).round(scale: 2);
    tax = (secFeeUsd * usdKrwRate).round(scale: 0);
  }

  final net = buy ? gross + fee : gross - fee - tax;
  return OrderAmount(grossAmount: gross, fee: fee, tax: tax, netAmount: net);
}

/// 이 예산([availableCash], 원)으로 매수할 수 있는 최대 수량.
/// 나눗셈 추정 뒤 [calculateOrderAmount]로 ±1주 오차를 보정한다 —
/// 원 단위 반올림 때문에 경계값에서 한 주 어긋날 수 있어서다.
int maxAffordableQuantity({
  required Decimal price,
  required String currency,
  required Decimal usdKrwRate,
  required Decimal availableCash,
}) {
  if (availableCash <= Decimal.zero || price <= Decimal.zero) return 0;

  final unitCostKrw = currency == 'USD' ? price * usdKrwRate : price;
  final estimatedCostPerShare = unitCostKrw * (Decimal.one + _feeRate);
  // 나눗셈은 무한소수가 될 수 있어 Rational.floor()로 바로 내린다.
  var qty = (availableCash / estimatedCostPerShare).floor().toInt();
  if (qty < 0) return 0;

  bool affordable(int q) =>
      calculateOrderAmount(
        side: 'BUY',
        quantity: q,
        price: price,
        currency: currency,
        usdKrwRate: usdKrwRate,
      ).netAmount <=
      availableCash;

  // 경계 보정은 설계상 ±1~2주 수준 — 최악의 비정상 입력에도 무한 반복하지 않도록
  // 횟수를 제한한다(웹과 동일).
  const maxSteps = 10;
  for (var steps = 0; qty > 0 && !affordable(qty) && steps < maxSteps; steps++) {
    qty--;
  }
  for (var steps = 0; affordable(qty + 1) && steps < maxSteps; steps++) {
    qty++;
  }
  return qty;
}
